src/backends/github/API.js

```diff
@@ -156,18 +156,35 @@ export default class API {
   }
 
   readFile(path, sha, branch = this.branch) {
-    const cache = sha ? LocalForage.getItem(`gh.${ sha }`) : Promise.resolve(null);
-    return cache.then((cached) => {
-      if (cached) { return cached; }
-
+    if (sha) {
+      return this.getBlob(sha);
+      // TODO: Should we fetch the file by path if the SHA is invalid?
+    } else {
       return this.request(`${ this.repoURL }/contents/${ path }`, {
         headers: { Accept: "application/vnd.github.VERSION.raw" },
         params: { ref: branch },
         cache: "no-store",
-      }).then((result) => {
-        if (sha) {
-          LocalForage.setItem(`gh.${ sha }`, result);
+      }).catch(error => {
+        if (hasIn(error, 'message.errors') && find(error.message.errors, { code:  "too_large" })) {
+          const dir = path.split('/').slice(0, -1).join('/');
+          return this.listFiles(dir)
+            .then(files => files.find(file => file.path === path))
+            // TODO: Should the case be handled where `file` is undefined? It should never happen because there is no pagination.
+            .then(file => getBlob(file.sha));
```

Comment:
Should the case be handled where `file` is undefined? It should never happen because there is no pagination.